libs/vkd3d/shared_fence.c

```diff
@@ -0,0 +1,149 @@
+#include "shared_fence.h"
+
+/* For now the implementation isn't cross-process.
+ * When fixing this, we just need to reimplement the blocking wait with an event handle,
+ * removing the need for the conditional variable, and use a win32 mutex instead of a critical section */
+
+#define MAX_WAIT_COUNT 100
+
+struct waiter_thread
+{
+    bool waiting, signaled;
+    HANDLE handle;
+    UINT64 value;
+};
+
+struct shared_fence
+{
+    DWORD pid;
+    UINT64 value;
+    CRITICAL_SECTION *cs;
+    CONDITION_VARIABLE *cv;
+    struct waiter_thread waiters[MAX_WAIT_COUNT];
+};
+
+HANDLE create_shared_fence(UINT64 initial_value)
+{
+    HANDLE mapping = CreateFileMappingW(INVALID_HANDLE_VALUE, NULL, PAGE_READWRITE | SEC_COMMIT, 0, sizeof(struct shared_fence), NULL);
+    struct shared_fence *fence = open_shared_fence(mapping);
+
+    fence->pid = GetCurrentProcessId();
+    fence->value = initial_value;
+    fence->cs = malloc(sizeof(*fence->cs));
```

Comment:
Is this a leak?

Reply:
closing conversation with new shared fence implementation